fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/ObjectAclFactory.java

```diff
@@ -0,0 +1,218 @@
+ */
+package edu.unc.lib.dl.acl.fcrepo4;
+
+import java.text.ParseException;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.jena.query.QueryExecution;
+import org.apache.jena.query.QuerySolution;
+import org.apache.jena.query.ResultSet;
+import org.apache.jena.rdf.model.RDFNode;
+import org.apache.jena.rdf.model.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import edu.unc.lib.dl.acl.service.PatronAccess;
+import edu.unc.lib.dl.acl.util.UserRole;
+import edu.unc.lib.dl.fedora.PID;
+import edu.unc.lib.dl.rdf.CdrAcl;
+import edu.unc.lib.dl.sparql.SparqlQueryService;
+import edu.unc.lib.dl.util.DateTimeUtil;
+
+/**
+ * Factory which provides access control details which are directly represented
+ * on objects, but does not take into account inheritance.
+ * 
+ * @author bbpennel
+ *
+ */
+public class ObjectAclFactory implements AclFactory {
+
+	private static final Logger log = LoggerFactory.getLogger(ObjectAclFactory.class);
+
+	private LoadingCache<String, List<Entry<String, String>>> objAclCache;
+	private long cacheTimeToLive;
+	private long cacheMaxSize;
+
+	private SparqlQueryService queryService;
+
+	private final List<String> roleUris;
+
+	public ObjectAclFactory() {
+		List<String> roleUris = new ArrayList<>(UserRole.values().length);
+
+		for (UserRole role : UserRole.values()) {
+			roleUris.add(role.getPropertyString());
+		}
+
+		this.roleUris = Collections.unmodifiableList(roleUris);
+	}
+
+	public void init() {
+		objAclCache = CacheBuilder.newBuilder()
+				.maximumSize(cacheMaxSize)
+				.expireAfterWrite(cacheTimeToLive, TimeUnit.MILLISECONDS)
+				.build(new ObjectAclCacheLoader());
+	}
+
+	@Override
+	public Map<String, Set<String>> getPrincipalRoles(PID pid) {
+		String pidString = pid.getRepositoryPath();
+		List<Entry<String, String>> objAcls = objAclCache.getUnchecked(pidString);
+
+		Map<String, Set<String>> result = objAcls.stream()
+				// Filter to only role assignments
+				.filter(p -> roleUris.contains(p.getKey()))
+				// Group up roles by principal
+				.collect(Collectors.groupingBy(
+						Entry<String, String>::getValue,
+						Collectors.mapping(Entry<String, String>::getKey, Collectors.toSet())
+					));
+
+		return result;
+	}
+
+	@Override
+	public PatronAccess getPatronAccess(PID pid) {
+		String pidString = pid.getRepositoryPath();
+		List<Entry<String, String>> objAcls = objAclCache.getUnchecked(pidString);
+
+		String patronPropertyUri = CdrAcl.patronAccess.getURI();
+		return objAcls.stream()
+				.filter(p -> patronPropertyUri.equals(p.getKey()))
+				.findFirst()
```

Comment:
It only makes sense to define one patron access setting on an object, and there was still work to be done with the response before returning so it couldn't be a terminal operator like matchAny, so I'm not sure if there are other ways to handle this?